xain_fl/config/schema.py

```diff
@@ -0,0 +1,389 @@
+"""This module provides helpers for reading and validating the TOML
+configuration.
+
+"""
+
+# Implementation notes:
+# =====================
+#
+# We use the `schema` library to validate the configuration provided
+# by the user. However, `schema` works mainly with dictionaries, which
+# are not particularly convenient for us:
+#
+# - we cannot generate documentation for all the dictionary keys
+# - the syntax for accessing values is quite verbose, especially if
+#   the dictionary has nested values
+#
+# Thus, although we use schemas to validate configurations (AI_SCHEMA,
+# SERVER_SCHEMA, and STORAGE_SCHEMA), we don't expose them. Instead,
+# we use them to dynamically generate classes where attributes are the
+# schema keys: AiConfig, ServerConfig, and StorageConfig. This
+# hackery happens in create_class_from_schema(). It is not
+# perfect. For instance, we cannot document the type of each
+# attribute. But it is arguably better than using plain dictionaries.
+
+from collections import namedtuple
+import ipaddress
+from typing import Any, Mapping, NamedTuple, Type, TypeVar
+import urllib
+
+import idna
+from schema import And, Optional, Or, Schema, SchemaError, Use
+import toml
+
+
+def error(key: str, description: str) -> str:
+    """Return an error message for the given configuration item and
+    description of the expected value type.
+
+    Args:
+
+        key (str): key of the configuration item
+        description (str): description of the expected type of value
+            for this configuration item
+    """
+    return f"Invalid `{key}`: value must be {description}"
```

Comment:
Please update this function after updating structlog message.

Reply:
@atymoshchuk sorry, I'm not sure what you mean here.